test/jdk/java/foreign/TestUpcall.java

```diff
@@ -30,23 +30,6 @@
  *
  * @run testng/othervm
  *   -Dforeign.restricted=permit
- *   -Djdk.internal.foreign.ProgrammableInvoker.USE_SPEC=false
```

Comment:
Here, in addition to reverting these changes, we should also test with all the new flavors of upcall specialization/intrinsics?

Reply:
I've tried to explain this in the PR body. Testing the current configuration already takes 15 minutes. Adding the new configurations would multiply the number of configurations by 4. This seems excessive. The full matrix can be tested periodically (maybe as part of a higher tier), but for regular testing runs we could only run the default configuration?

Reply:
Oops - I missed that part. On linux it doesn't take long to execute (1 min maybe) so I was surprised. But I understand it might be an issue in other platforms. That said, in the recent ResourceScope rewrite, executing using all modes has been helpful in shaking out issues, so I'm a bit worried to see this going completely. I'm pretty sure in a month I'll forget that I have to go in a manually edit to add the missing combo...

Reply:
One wacky idea: I assume these properties are not read-only - what if, the tests had an additional parameter which dictated the specific setup for generating upcall/downcalls? That way, we can decide how many stuff should be included when generating the various test combinations (and maybe just keep the main mode for standard executions).

Reply:
Yes, for such big patches it seems useful, but if the linker internals are untouched not so much. I can look a bit more at this and separate the full configuration out into a separate test group perhaps, e.g. we'd have like `run-test-jdk_foreign` and `run-test-jdk_foreign-full` (or with the default reversed).

Reply:
Isn't the additional parameter pretty much what we do now?

Reply:
Okay, I've added another test file called TestMatrix that contains all the configurations. This is not run by default, but could be run in isolation to shake out issues.

Reply:
I meant having a different data provider which tests every combo, by first setting JDK parameter, then creating downcall/upcall

Reply:
Hmm, would need to re-write the linker to query the property every time when linking, instead of initializing it once when the class is initialized. (I don't think it's possible to run tests in different forks with just TestNG). Seems like using a separate test file is more straight-forward? Maybe I'm misunderstanding but the result seems more or less the same.